# Mussel genomes assembly pipeline

import glob
import os.path
import re
from snakemake.utils import min_version

min_version("5.19.2")

configfile: "config/config.yaml"

container: config['sif_tools']

include: "rules/common.smk"

rule all:
    input: 
        expand("results/preprocessing/{sample}/{sample}_S1_L001_R1_001.fastq.gz",
            sample=config['samples']),
        expand("results/genomescope/{sample}/genomescope_res_{sample}/{sample}_preproc_summary.txt",
            sample=config['samples']),
        expand("results/supernova_assemblies/{sample}_{version}/DONE",
            sample=config['samples'], version=["v1", "v2"]),
        expand("results/fasta/{sample}_v3.pseudohap.fasta.gz",
            sample=config['samples']),
        expand("results/busco/{sample}_{version}_{db}/run_{db}/short_summary.txt",
            sample=config['samples'], version=["v1", "v2"],
            db=["metazoa_odb10", "mollusca_odb10"]),
        # expand("results/busco/{sample}_v3_{db}/run_{db}/short_summary.txt",
        #     sample=config['samples'],
        #     db=["metazoa_odb10", "mollusca_odb10"]),
        expand("results/busco/gallo_v0_{db}/run_{db}/short_summary.txt",
            db=["metazoa_odb10", "mollusca_odb10"]),
        # expand("results/kat/{sample}_{version}_spectra.pdf",
        #     sample=config['samples'], version=["v2", "v3"]),
        "results/stats/assembly_stats.csv"
        

include: "rules/preprocessing.smk"

include: "rules/genomescope.smk"

include: "rules/supernova.smk"

include: "rules/purge_dups.smk"

include: "rules/busco.smk"

include: "rules/kat.smk"

include: "rules/stats.smk"
